cpp/docs/DEVELOPER_GUIDE.md

```diff
@@ -0,0 +1,977 @@
+
+### View
+
+A view is a non-owning object that provides zero-copy access (possibly with slicing or offsets) data 
+owned by another object. Examples are column views and table views. 
+
+# Directory Structure and File Naming
+
+External/public libcudf APIs are grouped based on functionality into an appropriately titled 
+header file  in `cudf/cpp/include/cudf/`. For example, `cudf/cpp/include/cudf/copying.hpp` 
+contains the APIs for functions related to copying from one column to another. Note the  `.hpp` 
+file extension used to indicate a C++ header file.
+
+Header files should use the `#pragma once` include guard. 
+
+The naming of external API headers should be consistent with the name of the folder that contains 
+the source files that implement the API. For example, the implementation of the APIs found in
+`cudf/cpp/include/cudf/copying.hpp` are located in `cudf/src/copying`. Likewise, the unit tests for 
+the APIs reside in `cudf/tests/copying/`.
+
+Internal API headers containing `detail` namespace definitions that are used across translation 
+units inside libcudf should be placed in `include/cudf/detail`.
+
+## File extensions
+
+- `.hpp` : C++ header files
+- `.cpp` : C++ source files
+- `.cu`  : CUDA C++ source files
+- `.cuh` : Headers containing CUDA device code
+
+Only use `.cu` and `.cuh` if necessary. A good indicator is the inclusion of `__device__` and other
+symbols that are only recognized by `nvcc`. Another indicator is Thrust algorithm APIs with a device
+execution policy (always `rmm::exec_policy` in libcudf).
+
+## Code and Documentation Style and Formatting
+
+libcudf code uses [snake_case](https://en.wikipedia.org/wiki/Snake_case) for all names except in a 
+two cases: template parameters and unit tests and test case names may use Pascal case, aka 
+[UpperCamelCase](https://en.wikipedia.org/wiki/Camel_case). We do not use [Hungarian notation](https://en.wikipedia.org/wiki/Hungarian_notation), except sometimes when naming device data variables and their corresponding
+host copies. Private member variables are typically prefixed with an underscore.
+
+```c++
+template <typename IteratorType>
+void algorithm_function(int x, rmm::cuda_stream_view s, rmm::device_memory_resource* mr) 
+{
+  ...
+}
+
+class utility_class 
+{
+  ...
+ private:
+  int _rating{};
+  std::unique_ptr<cudf::column> _column{};
+}
+
+TYPED_TEST_CASE(RepeatTypedTestFixture, cudf::test::FixedWidthTypes);
+
+TYPED_TEST(RepeatTypedTestFixture, RepeatScalarCount)
+{
+  ...
+}
+```
+
+C++ formatting is enforced using `clang-format`. You should configure `clang-format` on your 
+machine to use the `cudf/cpp/.clang-format` configuration file, and run `clang-format` on all 
+changed code before committing it. The easiest way to do this is to configure your editor to 
+"format on save".
+
+Aspects of code style not discussed above and not automatically enforceable are typically caught
+during code review, or not enforced.
+
```

Comment:
Anything else we should include here?

- Usage of `static_cast` and `reinterpret_cast` instead of cast operator `(type)`
- Avoiding the comma operator https://github.com/rapidsai/cudf/pull/7090#discussion_r553330672
- Prefer using thrust or STL algorithms over for-loops when possible.

Reply:
I could just link the C++ core guidelines and Sean Parent's C++ seasoning talk and say "commit this to memory". :)

Reply:
Added a C++ Guidelines subsection. Thanks!